.github/workflows/pipeline.yml

```diff
@@ -0,0 +1,44 @@
+name: Pygmalion UI Front End - Build / Quality Control
+
+permissions:
+  contents: read # This is required for actions/checkout
+
+on:
+  workflow_dispatch:   # Allows you to run this workflow manually from the Actions tab
+
+jobs:
+  build:
+    name: Pipeline
+    runs-on: ubuntu-latest
+    steps:
```

Comment:
I think bundling everything under a single job is sub-optimal because:

- Everything will run sequentially, but theoretically after the initial install the lint/style/typecheck scripts are independent and could run in parallel
- If I make a mistake that breaks multiple checks (e.g. lint + style), I'll only see one of them until I fix it, and then get surprised by the other failure only after re-running CI

Do you think we could maybe get each step broken apart into a separate job, with the checkout/install step being a common dependency?

Reply:
> Sadly each job is hosted on its own runner/machine, so the dependencies steps had to be repeated in each one. Not as DRY as I would like, but I don't know of an alternative

You can mark the `build` job as a dependency in all the other jobs, and use artifacts to share data between them I believe. That's the proper way to do it so it only "builds" (installs) once and we can reuse everything across the parallel stages so things are sped up (at least on GitLab CI - I assume the same patterns can be used with GH Actions).

> Would we like some kind of notification of the results once completed - sent to email / matrix or something? historically i've had my GithubActions jobs send success/fail and a summary to Slack.
I assume there is similar integrations available for wherever we would like them sent

Probably not at the moment, everything builds so fast that I'll probably just see the results anyways whie reviewing the PRs. That might be a good idea for if this or the back-end project eventually starts to take long-ish on CI though.

Reply:
'Dependencies' job runs first, saving the dependencies cache

lint / style / typecheck are split to their own jobs. These use the shared dependencies cache, and run in parallel

Some steps have been moved to composite actions to keep the code DRY and easier to maintain
https://docs.github.com/en/actions/creating-actions/creating-a-composite-action

Reply:
Awesome! Thanks.